prow/pjutil/jobspec.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pjutil
+
+import (
+	"encoding/json"
+	"strconv"
+
+	"fmt"
+
+	"k8s.io/test-infra/prow/kube"
+)
+
+// JobSpec is the full downward API that we expose to
+// jobs that realize a ProwJob. We will provide this
+// data to jobs with environment variables in two ways:
+//  - the full spec, in serialized JSON in one variable
+//  - individual fields of the spec in their own variables
+type JobSpec struct {
```

Comment:
I don't feel strong about creating a spin-off of ProwJob. We already have one (albeit predates PJ) that we need to get rid of asap - see https://github.com/kubernetes/test-infra/issues/5216. Go doesn't help here, you won't be able to use any prow utilities if the need arises (I am a big fan of YAGNI but conversions between formats suck more). If we are going to do this, I would vote for using a PJ.

Reply:
My first pass was to just serialize the `ProwJobSpec`. That seems like sharing too much information -- the jobs do not need to care about retest commands, etc. I did not want to change the structure of a `ProwJob` internally as I did not know how much effort we would need to use to migrate the two versions when we upgraded the system. Is that easy? This is not only a good compromise but it allows us to expose the build ID and other ephemeral fields that we do not put into the `ProwJob` itself.